diff --git a/node_modules/@fuel-ts/providers/dist/index.d.ts b/node_modules/@fuel-ts/providers/dist/index.d.ts
index b174fbf..6d67c0b 100644
--- a/node_modules/@fuel-ts/providers/dist/index.d.ts
+++ b/node_modules/@fuel-ts/providers/dist/index.d.ts
@@ -2,7 +2,7 @@ import { BytesLike } from '@ethersproject/bytes';
 import { BigNumberish, BN } from '@fuel-ts/math';
 import { AbstractAddress, AbstractScriptRequest, ContractIdLike, AddressLike } from '@fuel-ts/interfaces';
 import { GraphQLClient } from 'graphql-request';
-import * as Dom from 'graphql-request/dist/types.dom';
+import * as Dom from 'graphql-request/build/esm/types.dom';
 import { Network } from '@ethersproject/networks';
 import { InputType, Input, OutputType, Output, TransactionType, TransactionScript, TransactionCreate, ReceiptCall, ReceiptReturn, ReceiptReturnData, ReceiptPanic, ReceiptRevert, ReceiptLog, ReceiptLogData, ReceiptTransfer, ReceiptTransferOut, ReceiptScriptResult, ReceiptMessageOut, Transaction } from '@fuel-ts/transactions';
 export { TransactionType } from '@fuel-ts/transactions';
